File: SystemUI/src/org/nameless/systemui/statusbar/policy/ForegroundActivityListener.java
/*
 * Copyright (C) 2024 The Nameless-AOSP Project
 * SPDX-License-Identifier: Apache-2.0
 */

package org.nameless.systemui.statusbar.policy;

import android.annotation.Nullable;
import android.content.Context;

import javax.inject.Inject;
import javax.inject.Singleton;

import org.nameless.view.AppFocusManager;
import org.nameless.view.IAppFocusObserver;

@Singleton
public class ForegroundActivityListener {

    private final AppFocusManager mAppFocusManager;

    private @Nullable String mFullscreenPackageName;
    private @Nullable String mFullscreenActivityName;

    private final IAppFocusObserver.Stub mAppFocusObserver = new IAppFocusObserver.Stub() {
        @Override
        public void onFullscreenFocusChanged(String packageName, String activityName) {
            mFullscreenPackageName = packageName;
            mFullscreenActivityName = activityName;
        }
    };

    @Inject
    public ForegroundActivityListener(Context context) {
        mAppFocusManager = context.getSystemService(AppFocusManager.class);
        mAppFocusManager.registerAppFocusObserver(mAppFocusObserver, true);
    }

    @Nullable
    public String getForegroundFullscreenPackageName() {
        return mFullscreenPackageName;
    }

    @Nullable
    public String getForegroundFullscreenActivityName() {
        return mFullscreenActivityName;
    }
}

File: SystemUI/src/org/nameless/systemui/statusbar/policy/ForegroundActivityListener.kt
/*
 * Copyright (C) 2024 The Nameless-AOSP Project
 * SPDX-License-Identifier: Apache-2.0
 */

package org.nameless.systemui.statusbar.policy

import android.content.Context

import javax.inject.Inject
import javax.inject.Singleton

import org.nameless.view.AppFocusManager
import org.nameless.view.IAppFocusObserver

@Singleton
class ForegroundActivityListener @Inject constructor(
    context: Context
) {

    private val appFocusManager = context.getSystemService(AppFocusManager::class.java)!!

    private var fullscreenPackageName = ""
    private var fullscreenActivityName = ""

    init {
        appFocusManager.registerAppFocusObserver(object : IAppFocusObserver.Stub() {
            override fun onFullscreenFocusChanged(packageName: String, activityName: String) {
                fullscreenPackageName = packageName
                fullscreenActivityName = packageName
            }
        }, true)
    }

    fun getTopPackageName() = fullscreenPackageName

    fun getTopActivityName() = fullscreenActivityName
}
